content/posts/2024-11-05-flang-new.md

```diff
@@ -0,0 +1,792 @@
+[runtime library](https://github.com/llvm/llvm-project/tree/a23260087db14032094d62dcf8be9be6f4e3d788/openmp). The final
+`.omp_outlined_debug__` version of `simple` is where the work actually happens.
+
+It includes labels like `omp.inner.for.body` but these are not specific to OpenMP.
+They are just normal LLVM IR labels.
+
+## Bringing OpenMP to MLIR
+
+Early in 2020, Kiran Chandramohan (Arm) [proposed](https://discourse.llvm.org/t/rfc-openmp-dialect-in-mlir/397) an MLIR dialect for OpenMP, with the end goal
+of Flang supporting OpenMP.
+
+> We (Arm) started the work for the OpenMP MLIR dialect because of Flang.
+> ... So, MLIR has an OpenMP dialect because of Flang.
+* Kiran Chandramohan
+
+This dialect would represent OpenMP specifically, unlike the generic LLVM IR you
+get from Clang.
+
+If you go back to the Fortran OpenMP example and compile it without OpenMP
+enabled, you get this MLIR:
+
+```mlir
+module attributes {<...>} {
+  func.func @_QPsimple(<...> {
+    %1:2 = hlfir.declare %arg0 <...> {uniq_name = "_QFsimpleEn"} : <...>
+    %3:2 = hlfir.declare %2 <...> {uniq_name = "_QFsimpleEi"} : <...>
+    %10:2 = hlfir.declare %arg1(%9) <...> {uniq_name = "_QFsimpleEa"} : <...>
+    %17:2 = hlfir.declare %arg2(%16) <...> {uniq_name = "_QFsimpleEb"} : <...>
+    %22:2 = fir.do_loop <...> {
+      <...>
+      hlfir.assign %34 to %37 : f32, !fir.ref<f32>
+    }
+    fir.store %22#1 to %3#1 : !fir.ref<i32>
+    return
+  }
+}
+```
+(output heavily edited for readability)
+
+Notice that the `DO` loop has been converted into `for.do_loop`. Now enable
+OpenMP and compile again:
+
+```mlir
+module attributes {<...>} {
+  func.func @_QPsimple(<...> {
+    %1:2 = hlfir.declare %arg0 <...> {uniq_name = "_QFsimpleEn"} : <...>
+    %10:2 = hlfir.declare %arg1(%9) <...> {uniq_name = "_QFsimpleEa"} : <...>
+    %17:2 = hlfir.declare %arg2(%16) <...> {uniq_name = "_QFsimpleEb"} : <...>
+    omp.parallel {
+      %19:2 = hlfir.declare %18 {uniq_name = "_QFsimpleEi"} : <...>
+      omp.wsloop {
+        omp.loop_nest (%arg3) : i32 = (%c2_i32) to (%20) inclusive step (%c1_i32) {
+          hlfir.assign %32 to %35 : f32, !fir.ref<f32>
+          omp.yield
+        }
+      }
+      omp.terminator
+    }
+    return
+  }
+}
+```
+(output heavily edited for readability)
+
+You will see that instead of `for.do_loop` you have `omp.parallel`. `omp` is
+MLIR's [OpenMP dialect](https://mlir.llvm.org/docs/Dialects/OpenMPDialect/).
+This is a higher level and much more literal translation of the
+`parallel` directive than Clang produced.
+
+As the `omp` dialect is specifically made for OpenMP, it can represent
+it in a much more natural manner. This makes understanding the code and writing
+optimisations much easier.
+
+Of course Flang needs to produce LLVM IR eventually, and it reused Clang's
+`OpenMPIRBuilder` library for this. Below is the LLVM IR produced from the MLIR
+above:
+
+```mlir
+define void @simple_(<...> {
```

Comment:
nit: this is a matter of preference (so feel free to ignore it), but the omitted closing parentheses in this snippet makes it harder for me to parse it.

Reply:
Yeah I've cleaned up a few of these. It'll help because there is formatting for MLIR so it'll colour matched sets nicely.